rclcpp/src/rclcpp/time_source.cpp

```diff
@@ -0,0 +1,223 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <utility>
+
+#include "builtin_interfaces/msg/time.hpp"
+
+#include "rcl/time.h"
+
+#include "rcutils/logging_macros.h"
+
+#include "rclcpp/clock.hpp"
+#include "rclcpp/exceptions.hpp"
+#include "rclcpp/node.hpp"
+#include "rclcpp/time.hpp"
+#include "rclcpp/time_source.hpp"
+
+
+namespace rclcpp
+{
+
+TimeSource::TimeSource(std::shared_ptr<rclcpp::node::Node> node)
+: ros_time_valid_(false)
+{
+  this->attachNode(node);
+}
+
+TimeSource::TimeSource()
+: ros_time_valid_(false)
+{
+}
+
+void TimeSource::attachNode(std::shared_ptr<rclcpp::node::Node> node)
+{
+  node_ = node;
+  // TODO(tfoote): Update QOS
+  clock_subscription_ = node_->create_subscription<builtin_interfaces::msg::Time>(
+    "clock", std::bind(&TimeSource::clock_cb, this, std::placeholders::_1),
+    rmw_qos_profile_default);
+
+  parameter_client_ = std::make_shared<rclcpp::parameter_client::AsyncParametersClient>(node);
+  parameter_subscription_ =
+    parameter_client_->on_parameter_event(std::bind(&TimeSource::on_parameter_event,
+      this, std::placeholders::_1));
+}
+
+void TimeSource::detachNode()
+{
+  this->ros_time_valid_ = false;
+  clock_subscription_.reset();
+  parameter_client_.reset();
+  node_.reset();
+  disableROSTime();
+}
+
+void TimeSource::attachClock(std::shared_ptr<rclcpp::Clock> clock)
+{
+  std::lock_guard<std::mutex> guard(clock_list_lock_);
+  associated_clocks_.push_back(clock);
+  // TODO(tfoote) cache state and set clock here
+}
+
+void TimeSource::detachClock(std::shared_ptr<rclcpp::Clock> clock)
+{
+  std::lock_guard<std::mutex> guard(clock_list_lock_);
+  auto result = std::find(associated_clocks_.begin(), associated_clocks_.end(), clock);
+  if (result != associated_clocks_.end()) {
+    associated_clocks_.erase(result);
+  } else {
+    RCUTILS_LOG_ERROR("Failed to remove clock");
+  }
+}
+
+TimeSource::~TimeSource()
+{
+  if (node_) {
+    this->detachNode();
+  }
+}
+
+void TimeSource::setClock(const builtin_interfaces::msg::Time::SharedPtr msg,
+  std::shared_ptr<rclcpp::Clock> clock)
+{
+  // TODO(tfoote) Move enabledROS time logic here.
+
+  rclcpp::Time msg_time = rclcpp::Time(*msg);
+  rclcpp::Time now = clock->now();
+  auto diff = now - msg_time;
+  rclcpp::TimeJump jump;
+  jump.delta_.nanoseconds = diff.nanoseconds();
+  // TODO(tfoote) fill in time source change on jump
+
+
+  // TODO(tfoote) potential race condition with target going out of scope
```

Comment:
What is `target` in this context? `clock` is a shared pointer and `jump` is local. What could go out of scope?

Reply:
I moved this to state the required lifetime for the validity of the callback functions compared to the handler.